### PR TITLE
Redirection refactored

### LINKED ISSUE #105: Goldencobra::Redirector is undefined/not-working for certain edge cases
This test:
https://github.com/ikuseiGmbH/Goldencobra/blob/69a1ab26b789282abc0a71a47675c233618d994c/test/dummy/spec/models/redirector_spec.rb#L193-L200

does not work. This is why:

https://github.com/ikuseiGmbH/Goldencobra/blob/69a1ab26b789282abc0a71a47675c233618d994c/app/models/goldencobra/redirector.rb#L70-L104

In line 79, we try to match the request with existing redirector records, but only match with the given request-path, ignoring any url params. The existing redirector record does have a parameter at the end, though. So it cannot match.

Our current idea is, to expand the `Goldencobra::Redirector` model with new url-parameter attributes. The matching would then again be done by the param-free url path, and additionally taking any given url-params into consideration.

### PR DESCRIPTION
**Pull request by marcometz**
URL params in source_url are now required by default if present and handled by redirector

Fixes #105

### CODE REVIEW

File: app/models/goldencobra/redirector.rb [hunk 1]
```diff
@@ -74,26 +74,18 @@ def self.get_by_request(request_original_url)
...
         request_path = "#{uri.scheme}://#{uri.host}#{uri.path}"
-        redirects = Goldencobra::Redirector.active.where(source_url: request_path)
-        if redirects.any?
-          #if multiple redirectors found, select the first
-          redirect = redirects.first
-          redirecter_source_uri = Addressable::URI.parse(redirect.source_url)
+        regexp = "#{request_path}[?$]*.*"
+        redirects = Goldencobra::Redirector.active.where("source_url RLIKE '#{regexp}'")
+
+        matching_redirection = redirects.select{|a| compare_url_params(url_params_from_url(a.source_url), url_params_from_url(request_original_url)) }.first
```
Comment (donni106) [inline app/models/goldencobra/redirector.rb]: [linting] spaces `.select { ... }`
und sehr gerne `|r|` statt `|a|`

File: test/dummy/Guardfile [hunk 2]
```diff
@@ -23,8 +23,8 @@ guard 'rspec', cmd: "bundle exec rspec" do
...
-guard 'cucumber', cmd: "bundle exec cucumber" do
-  watch(%r{^features/.+\.feature$})
-  watch(%r{^features/support/.+$})          { 'features' }
-  watch(%r{^features/step_definitions/(.+)_steps\.rb$}) { |m| Dir[File.join("**/#{m[1]}.feature")][0] || 'features' }
-end
+# guard 'cucumber', cmd: "bundle exec cucumber" do
+#   watch(%r{^features/.+\.feature$})
+#   watch(%r{^features/support/.+$})          { 'features' }
+#   watch(%r{^features/step_definitions/(.+)_steps\.rb$}) { |m| Dir[File.join("**/#{m[1]}.feature")][0] || 'features' }
+# end
```
Comment (donni106) [inline test/dummy/Guardfile]: Soll das hier so committed werden?

File: app/models/goldencobra/redirector.rb [hunk 3]
```diff
@@ -103,6 +95,22 @@ def self.get_by_request(request_original_url)
       end
     end
 
+    # Bestimmung, ob alle Parameter in source_ulr
```
[the remark this answers is not in the document]
Comment (marcometz) [reply app/models/goldencobra/redirector.rb]: korrigiert

File: app/models/goldencobra/redirector.rb [hunk 4]
```diff
@@ -103,6 +95,22 @@ def self.get_by_request(request_original_url)
...
+    # auch in der Request_url enthalten sind
```
[the remark this answers is not in the document]
Comment (marcometz) [reply app/models/goldencobra/redirector.rb]: korrigiert

File: app/models/goldencobra/redirector.rb [hunk 5]
```diff
@@ -103,6 +95,22 @@ def self.get_by_request(request_original_url)
...
+    #
+    # @param [Array] source_params list of hashes
+    # @param [Array] request_params list of hashes
+    #
+    # @return [Boolean>] True, wenn alle Parameter in source_ulr auch in der Request_url enthalten sind
```
[the remark this answers is not in the document]
Comment (marcometz) [reply app/models/goldencobra/redirector.rb]: korrigiert

File: test/dummy/Guardfile (hunk 2 above)
Comment (marcometz) [reply test/dummy/Guardfile]: ja, die funktionieren eh alle nicht :-(
